### PR TITLE
[Java] Discriminator lookup should not be case insensitive by default

### PR DESCRIPTION
**Pull request by sebastien-rosset**
1. Case sensitivity

The OpenAPI specification explicitly states field names are case sensitive:
> All field names in the specification are case sensitive. This includes all fields that are used as keys in a map, except where explicitly noted that keys are case insensitive.

That includes discriminator values. The discriminator value lookup must be case sensitive by default. Adding a property to support case insensitive lookup.

2. Incorrect class name in generated discriminator lookup.

In the generated JSON.java file, there is a lookup function that takes a string as input and returns the name of the Java class. The map incorrectly uses the name of the Java class to register the mapping in classByDiscriminatorValue. This isn't a problem if the OAS document has a name that does not require sanitization (e.g. "Animal"), but for example "animal" would be converted to "Animal" when creating a Java class.

### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) before.
- [x] Run the shell script(s) under `./bin/` (or Windows batch scripts under`.\bin\windows`) to update Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit, and these must match the expectations made by your contribution. You only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the code or mustache templates for a language (`{LANG}`) (e.g. php, ruby, python, etc).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.3.x`, `5.0.x`. Default: `master`.
- [x] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### CODE REVIEW

**Comment by sebastien-rosset** [comment]
@wing328

File: samples/client/petstore/java/feign/src/main/java/org/openapitools/client/model/Animal.java [hunk 1]
```diff
@@ -43,7 +43,7 @@
 
 public class Animal {
   public static final String JSON_PROPERTY_CLASS_NAME = "className";
-  private String className;
+  protected String className;
```
Comment (sebastien-rosset) [inline samples/client/petstore/java/feign/src/main/java/org/openapitools/client/model/Animal.java:46]: @wing328 , what causes the change from private to protected?

**Comment by sebastien-rosset** [comment]
@bbdouglas  @sreeshas @jfiala @lukoyanov  @cbornet @jeff9finger  @karismann  @Zomzog @lwlee2608  @bkabrda

**Comment by wing328** [comment]
@sebastien-rosset thanks for the PR. I would consider this a bug fix so eventually, we may want to remove this option (maybe in 5.x) to ensure the correct behavior moving forward.

**Comment by sebastien-rosset** [comment]
> @sebastien-rosset thanks for the PR. I would consider this a bug fix so eventually, we may want to remove this option (maybe in 5.x) to ensure the correct behavior moving forward.

Agreed.

**Pull request merged**